framework/logging.cpp

```diff
@@ -1435,26 +1418,59 @@ static void logging_format_data(DataType type, std::string_view description, con
         break;
     }
 
-    buffer +=
-            stdprintf("%sdescription: '%.*s'\n"
-                      "%stype:        %s\n"
-                      "%soffset:      [ %td, %td ]\n"
-                      "%saddress:     '%p'\n",
-                      spaces.c_str(), int(description.size()), description.data(),
-                      spaces.c_str(), typeName,
-                      spaces.c_str(), alignedOffset, offset - alignedOffset,
-                      spaces.c_str(), data1 + offset);
-    if (uint64_t physaddr = retrieve_physical_address(data1 + offset)) {
-        // 2^48-1 requires 12 hex digits
-        buffer +=  stdprintf("%sphysical:    '%#012" PRIx64 "'\n",
-                             spaces.c_str(), physaddr);
-    }
-    buffer += stdprintf("%sactual:      '0x%s'\n"
-                        "%sexpected:    '0x%s'\n"
-                        "%smask:        '0x%s'\n",
-                        spaces.c_str(), format_single_type(type, typeSize, data1 + alignedOffset, true).c_str(),
-                        spaces.c_str(), format_single_type(type, typeSize, data2 + alignedOffset, true).c_str(),
-                        spaces.c_str(), format_single_type(type, typeSize, xormask, false).c_str());
+    auto formatAddresses = [&spaces](const uint8_t *ptr) {
+        std::string result = stdprintf("%saddress:     '%p'\n", spaces.c_str(), ptr);
+        if (uint64_t physaddr = retrieve_physical_address(ptr)) {
```

Comment:
I wouldn't do that either, if the variable weren't declared in that line. It's impossible to get this wrong if the type is there, because `uint64_t physaddr == something` doesn't compile.